_posts/2020-03-02-tomcat-thread-model.md

```diff
@@ -0,0 +1,96 @@
+
+### Test
+
+Dla porównania można wykonać prosty test na dwóch rodzajach connectorów. W teście klient raz na sekundę wysyła żądanie do serwera. Łącznie jeden klient wysyła 30 żądań. Uruchamiamy 20 klientów w ciągu 2 sekund. Jednocześnie monitorujemy rozmiar puli wątków.
+
+Tak wygląda wykres czasów odpowiedzi dla connectora blokującego:
+
+![gatling-bio.png](/assets/img/posts/2020-03-02-tomcat-thread-model/gatling-bio.png)
+
+a tak dla connectora nieblokującego:
+
+![gatling-nio.png](/assets/img/posts/2020-03-02-tomcat-thread-model/gatling-nio.png) 
+
+Nie widać tutaj jakiejś specjalnej różnicy między działaniem connectora bio i nio. Zupełnie inaczej wyglądają natomiast wykresy liczby zajętych wątków w puli. Dla connectora blokującego liczba zajętych wątków podczas trwania testu przekracza 20. Czyli jest zgodna z liczbą klientów wysyłających żądania do serwera. Połączenia keep-alive mają timeout równy 5 sekund więc klient utrzymuje cały czas jedno połączenie do wszystkich żądań:
+
+![threads-bio.png](/assets/img/posts/2020-03-02-tomcat-thread-model/threads-bio.png)
+
+Dla connectora nieblokującego żądania są obsługiwane w większości przez jeden wątek:
+
+![threads-nio.png](/assets/img/posts/2020-03-02-tomcat-thread-model/threads-nio.png)
+
+Test pokazuje, że przy takim modelu przetwarzania żądań za pomocą connectora nieblokującego możemy uzyskać sporą oszczędność zasobów i lepiej zutylizować serwer.
+
+### Użycie mod_proxy zmienia sytuację
+
+Warto zwrócić uwagę na to, że takie wyniki osiągamy w przypadku bezpośredniego połączenia pomiędzy klientem a serwerem Tomcat. Zdarza się jednak, że pomiędzy serwerem Tomcat a klientem mamy jeszcze warstwę pośrednią np. w postaci serwera Apache httpd i modułu mod_proxy. Domyślnie mod_proxy nie ma włączonej opcji keep-alive więc wszystkie połączenia do backendu są zamykane po obsłużeniu pojedynczego żądania. 
+
+Test został przeprowadzony za pomocą narzędzia [Gatling](https://gatling.io/). Więcej o samym narzędziu można przeczytać w osobnym [artykule na naszym blogu](https://blog.consdata.tech/2017/08/01/gatling.html).
+
+## Próba rozwiązania problemu
+
+Podsumowując: Wstępem do rozważań nad konfiguracją connectorów był problem z długimi czasami jakie wątki Tomcata spędzały na komunikacji z systemem autoryzacji. Czy zmiana connectora na nieblokujący pomogłaby w tej sytuacji? Niestety nie. Connector nieblokujący nie wiąże wątku z połączeniem, ale nadal wiąże go z pojedynczym żądaniem. W tym przypadku nadal wątki będą blokowane na długim wywołaniu systemu autoryzacji. Choć zastosowanie connectora nio z pewnością zaoszczędziłoby część zasobów maszyny to w tym przypadku należałoby raczej rozważyć użycie jakiejś implementacji circuit breaker'a lub też zmienić model obsługi żądań na bardziej reaktywny (np. [webflux](https://docs.spring.io/spring/docs/current/spring-framework-reference/web-reactive.html)).
```

Comment:
przecinek przez "to":
Choć zastosowanie connectora nio z pewnością zaoszczędziłoby część zasobów maszyny to w tym przypadku
==>
Choć zastosowanie connectora nio z pewnością zaoszczędziłoby część zasobów maszyny, to w tym przypadku

Reply:
Poprawione.